### PR TITLE
[Object Detection] Use functional subclass paradigm for RetinaNet

### PR DESCRIPTION
**Pull request by ianstenbit**

### CODE REVIEW

File: keras_cv/models/object_detection/retina_net/retina_net.py [hunk 1]
```diff
@@ -308,7 +314,7 @@ def compile(
         super().compile(loss=losses, **kwargs)
 
     def compute_loss(self, images, gt_boxes, gt_classes, training):
-        box_pred, cls_pred = self._forward(images, training=training)
+        box_pred, cls_pred = self.call(images, training=training)
```
Comment (tanzhenyu) [inline keras_cv/models/object_detection/retina_net/retina_net.py:317]: if you use `call` here, would the validation loss be out of order (compared to the train loss)?

File: keras_cv/models/object_detection/retina_net/retina_net.py [hunk 2]
```diff
@@ -411,6 +417,19 @@ def test_step(self, data):
 
         return self.compute_metrics(x, {}, {}, sample_weight={})
 
+    def get_config(self):
+        return {
+            "classes": self.classes,
+            "bounding_box_format": self.bounding_box_format,
+            "backbone": self.backbone,
```
Comment (tanzhenyu) [inline keras_cv/models/object_detection/retina_net/retina_net.py:424]: don't we need to call `self.backbone.get_config` here?

**Review by tanzhenyu — COMMENTED** [review]
Thanks for the PR!

File: keras_cv/models/object_detection/retina_net/retina_net.py (hunk 1 above)
Comment (tanzhenyu) [reply keras_cv/models/object_detection/retina_net/retina_net.py:317]: (as the validation_data would use `test_step` which calls `call` with training=False, so the decoded box is used to compute validation loss?

File: keras_cv/models/object_detection/retina_net/retina_net.py (hunk 2 above)
Comment (ianstenbit) [reply keras_cv/models/object_detection/retina_net/retina_net.py:425]: Since the backbone is a Keras concept, config can be loaded recursively (same thing we do in RetinaNetLabelEncoder with the AnchorGenerator).

But we were missing serialization registration -- Done

File: keras_cv/models/object_detection/retina_net/retina_net.py (hunk 1 above)
Comment (ianstenbit) [reply keras_cv/models/object_detection/retina_net/retina_net.py:318]: Good catch -- I think this should be calling `super().call` (which should be identical to `_forward`)

I updated it to call super().call and the validation loss looks correct.
Without the change, the validation loss is super high (as one might expect)

**Review by tanzhenyu — APPROVED** [review]
Please verify the result is >32 mAP after training 30 epochs

**Comment by ianstenbit** [comment]
> Please verify the result is >32 mAP after training 30 epochs

SG -- training now

**Comment by tanzhenyu** [comment]
> > Please verify the result is >32 mAP after training 30 epochs
> 
> SG -- training now

FWIW, on TPU v3-8 it's training at 175ms/step, and >32 mAP. Let me know if you need me to share the training loss / tensorboard

**Comment by ianstenbit** [comment]
> > > Please verify the result is >32 mAP after training 30 epochs
> > 
> > 
> > SG -- training now
> 
> FWIW, on TPU v3-8 it's training at 175ms/step, and >32 mAP. Let me know if you need me to share the training loss / tensorboard

On a TPUv3-8 with this change I am seeing 175ms/step and an MAP of 30 after 20 epochs. If it hits 32mAP by 30 epochs, I will merge this PR.
